### PR TITLE
Make `compile_options` an opaque object for ABI compatibility

### PR DESCRIPTION
**Pull request by umangyadav**
Depends on PR https://github.com/ROCmSoftwarePlatform/AMDMIGraphX/pull/953 

Review by commits.  1. `file_options` commit has changes from PR953. 2. second commit has changes related to this pr.

### CODE REVIEW

**Review by kahmed10 — APPROVED** [review]

File: src/api/api.cpp [hunk 1]
```diff
@@ -1,4 +1,3 @@
-#include <migraphx/operation.hpp>
```
Comment (umangyadav) [inline src/api/api.cpp:1]: @scxiao @pfultz2 @causten Header is removed.

File: src/api/include/migraphx/migraphx.hpp [hunk 2]
```diff
@@ -494,6 +494,25 @@ struct module
...
 };
 
+struct compile_options : MIGRAPHX_HANDLE_BASE(compile_options)
+{
+    compile_options() { this->make_handle(&migraphx_compile_options_create); }
+
+    compile_options(migraphx_compile_options* p, own) { this->set_handle(p, own()); }
+
+    // set offload_copy
+    void set_offload_copy(bool value)
```
Comment (pfultz2) [inline src/api/include/migraphx/migraphx.hpp:504]: Make `value` default to `true`.

File: src/api/include/migraphx/migraphx.hpp [hunk 3]
```diff
@@ -494,6 +494,25 @@ struct module
...
+    {
+        call(&migraphx_compile_options_set_offload_copy, this->get_handle_ptr(), value);
+    }
+
+    // set fast_math
+    void set_fast_math(bool value)
```
Comment (pfultz2) [inline src/api/include/migraphx/migraphx.hpp:510]: Make `value` default to `true`.

File: src/api/include/migraphx/migraphx.hpp (hunk 2 above)
Comment (umangyadav) [reply src/api/include/migraphx/migraphx.hpp]: Done

File: src/api/include/migraphx/migraphx.hpp (hunk 3 above)
Comment (umangyadav) [reply src/api/include/migraphx/migraphx.hpp]: Done

File: test/api/test_gpu.cpp [hunk 4]
```diff
@@ -7,8 +7,8 @@ TEST_CASE(load_and_run)
 {
     auto p             = migraphx::parse_onnx("conv_relu_maxpool_test.onnx");
     auto shapes_before = p.get_output_shapes();
-    migraphx_compile_options options;
-    options.offload_copy = true;
+    migraphx::compile_options options;
+    options.set_offload_copy(true);
```
Comment (pfultz2) [inline test/api/test_gpu.cpp:11]: Just do `options.set_offload_copy()`.

Comment (umangyadav) [reply test/api/test_gpu.cpp]: Done

File: tools/api/migraphx.h [hunk 5]
```diff
@@ -41,19 +41,6 @@ typedef enum {
 } migraphx_shape_datatype_t;
 #undef MIGRAPHX_SHAPE_GENERATE_ENUM_TYPES
 
-/// Options to be passed when compiling
-typedef struct
-{
-    /// For targets with offloaded memory(such as the gpu), this will insert
-    /// instructions during compilation to copy the input parameters to the
-    /// offloaded memory and to copy the final result from the offloaded
-    /// memory back to main memory.
```
[the remark this answers is not in the document]
Comment (umangyadav) [reply tools/api/migraphx.h:50]: Done

File: test/api/test_gpu.cpp [hunk 6]
```diff
@@ -1,14 +1,26 @@
 #include <numeric>
 #include <migraphx/migraphx.h>
 #include <migraphx/migraphx.hpp>
+#include <migraphx/compile_options.hpp>
 #include "test.hpp"
 
+TEST_CASE(compile_options_api_test)
```
Comment (pfultz2) [inline test/api/test_gpu.cpp:7]: Can you move this test to a different C++ file? This way we dont need to change everything to use `migraphx::api::compile_options` in this test file.

Comment (umangyadav) [reply test/api/test_gpu.cpp]: Done.

**Review by scxiao — APPROVED** [review]

**Review by pfultz2 — APPROVED** [review]

**Pull request merged**